compiler/shared_ast/typing.ml

```diff
@@ -225,73 +225,48 @@ let lit_type (type a) (lit : a A.glit) : naked_typ =
-  let rt = UnionFind.make (TLit TRat, pos) in
-  let mt = UnionFind.make (TLit TMoney, pos) in
-  let dut = UnionFind.make (TLit TDuration, pos) in
-  let dat = UnionFind.make (TLit TDate, pos) in
-  let any = UnionFind.make (TAny (Any.fresh ()), pos) in
-  let array_any = UnionFind.make (TArray any, pos) in
-  let any2 = UnionFind.make (TAny (Any.fresh ()), pos) in
-  let array_any2 = UnionFind.make (TArray any2, pos) in
-  let arr x y = UnionFind.make (TArrow (x, y), pos) in
-  match Marked.unmark op with
-  | A.Ternop A.Fold ->
-    arr (arr any2 (arr any any2)) (arr any2 (arr array_any any2))
-  | A.Binop (A.And | A.Or | A.Xor) -> arr bt (arr bt bt)
-  | A.Binop (A.Add KInt | A.Sub KInt | A.Mult KInt | A.Div KInt) ->
-    arr it (arr it it)
-  | A.Binop (A.Add KRat | A.Sub KRat | A.Mult KRat | A.Div KRat) ->
-    arr rt (arr rt rt)
-  | A.Binop (A.Add KMoney | A.Sub KMoney) -> arr mt (arr mt mt)
-  | A.Binop (A.Add KDuration | A.Sub KDuration) -> arr dut (arr dut dut)
-  | A.Binop (A.Sub KDate) -> arr dat (arr dat dut)
-  | A.Binop (A.Add KDate) -> arr dat (arr dut dat)
-  | A.Binop (A.Mult KDuration) -> arr dut (arr it dut)
-  | A.Binop (A.Div KMoney) -> arr mt (arr mt rt)
-  | A.Binop (A.Mult KMoney) -> arr mt (arr rt mt)
-  | A.Binop (A.Lt KInt | A.Lte KInt | A.Gt KInt | A.Gte KInt) ->
-    arr it (arr it bt)
-  | A.Binop (A.Lt KRat | A.Lte KRat | A.Gt KRat | A.Gte KRat) ->
-    arr rt (arr rt bt)
-  | A.Binop (A.Lt KMoney | A.Lte KMoney | A.Gt KMoney | A.Gte KMoney) ->
-    arr mt (arr mt bt)
-  | A.Binop (A.Lt KDate | A.Lte KDate | A.Gt KDate | A.Gte KDate) ->
-    arr dat (arr dat bt)
-  | A.Binop (A.Lt KDuration | A.Lte KDuration | A.Gt KDuration | A.Gte KDuration)
-    ->
-    arr dut (arr dut bt)
-  | A.Binop (A.Eq | A.Neq) -> arr any (arr any bt)
-  | A.Binop A.Map -> arr (arr any any2) (arr array_any array_any2)
-  | A.Binop A.Filter -> arr (arr any bt) (arr array_any array_any)
-  | A.Binop A.Concat -> arr array_any (arr array_any array_any)
-  | A.Unop (A.Minus KInt) -> arr it it
-  | A.Unop (A.Minus KRat) -> arr rt rt
-  | A.Unop (A.Minus KMoney) -> arr mt mt
-  | A.Unop (A.Minus KDuration) -> arr dut dut
-  | A.Unop A.Not -> arr bt bt
-  | A.Unop (A.Log (A.PosRecordIfTrueBool, _)) -> arr bt bt
-  | A.Unop (A.Log _) -> arr any any
-  | A.Unop A.Length -> arr array_any it
-  | A.Unop A.GetDay -> arr dat it
-  | A.Unop A.GetMonth -> arr dat it
-  | A.Unop A.GetYear -> arr dat it
-  | A.Unop A.FirstDayOfMonth -> arr dat dat
-  | A.Unop A.LastDayOfMonth -> arr dat dat
-  | A.Unop A.RoundMoney -> arr mt mt
-  | A.Unop A.RoundDecimal -> arr rt rt
-  | A.Unop A.IntToRat -> arr it rt
-  | A.Unop A.MoneyToRat -> arr mt rt
-  | A.Unop A.RatToMoney -> arr rt mt
-  | Binop (Mult KDate) | Binop (Div (KDate | KDuration)) | Unop (Minus KDate) ->
-    Errors.raise_spanned_error pos "This operator is not available!"
+  let any = lazy (UnionFind.make (TAny (Any.fresh ()), pos)) in
+  let any2 = lazy (UnionFind.make (TAny (Any.fresh ()), pos)) in
```

Comment:
What is this lazy devilish sorcery? Are we in Haskell now? :)

Reply:
:rofl: 
The original version was creating a lot of `UnionFind.t` that never got used, for every operator met in the code.
- Inlining them would have made the code verbose and less readable
- On the other hand, turning them into functions wouldn't have worked since we want unicity of the `any` and `any2` variables.

sooo... making them lazy just seemed the most lightweight way to retain the original behaviour, but avoid the unnecessary creations :shrug: ?

Reply:
Yeah this makes sense. Not sure if this optimizations saves a lot of things but it sounds like a legit use of `lazy` :)